File: app/timeasy/lib/tools/weekly_statistics.dart
class WeeklyStatistics {
  Map<int, WeeklyStatisticsEntry> _entries = new Map();

  WeeklyStatisticsEntry? getEntryForWeekDay(int weekDay) {
    return _entries[weekDay];
  }

  void addEntryForWeekDay(int weekDay, WeeklyStatisticsEntry entry) {
    _entries[weekDay] = entry;
  }

  int getSumInSeconds() {
    var _sum = 0;
    _entries.forEach((weekday, entry) => _sum += entry.seconds);
    return _sum;
  }
}

class WeeklyStatisticsEntry {
  final DateTime date;
  int seconds = 0;

  WeeklyStatisticsEntry(this.date) {}

  double getMinutes() {
    return seconds / 60;
  }

  double getHours() {
    return getMinutes() / 60;
  }
}

File: server/src/pkg/domain/model/weekly_statistics.go
package model

import (
	"github.com/gofrs/uuid"
	"time"
)

func NewWeeklyStatistics() *WeeklyStatistics {
	return &WeeklyStatistics{
		entries: make(map[time.Weekday]*WeeklyStatisticsEntry),
	}
}

type WeeklyStatistics struct {
	entries map[time.Weekday]*WeeklyStatisticsEntry
}

func (stats *WeeklyStatistics) AddEntryForWeekday(weekday time.Weekday, entry *WeeklyStatisticsEntry) {
	stats.entries[weekday] = entry
}

func (stats *WeeklyStatistics) GetEntryForWeekday(weekday time.Weekday) *WeeklyStatisticsEntry {
	return stats.entries[weekday]
}

func (stats *WeeklyStatistics) GetSecondsForWeekday(weekday time.Weekday) int {
	entry := stats.GetEntryForWeekday(weekday)
	if entry == nil {
		return 0
	}
	return entry.Seconds
}

func (stats *WeeklyStatistics) GetSumInSeconds() int {
	var sum int
	for _, entry := range stats.entries {
		sum += entry.Seconds
	}
	return sum
}

func NewWeeklyStatisticsEntry(date time.Time) *WeeklyStatisticsEntry {
	return &WeeklyStatisticsEntry{
		Date: date,
	}
}

type WeeklyStatisticsEntry struct {
	Date            time.Time
	Seconds         int
	TimesPerProject map[uuid.UUID]*TimePerProject
}

type TimePerProject struct {
	ProjectId uuid.UUID
	Seconds   int
}

func (entry *WeeklyStatisticsEntry) GetMinutes() float32 {
	return float32(entry.Seconds) / 60
}

func (entry *WeeklyStatisticsEntry) GetHours() float32 {
	return entry.GetMinutes() / 60
}

func (entry *WeeklyStatisticsEntry) AddSecondsForProject(projectId uuid.UUID, seconds int) {
	if entry.TimesPerProject == nil {
		entry.TimesPerProject = make(map[uuid.UUID]*TimePerProject)
	}
	timeOfProject := entry.TimesPerProject[projectId]
	if timeOfProject == nil {
		timeOfProject = &TimePerProject{
			ProjectId: projectId,
			Seconds:   0,
		}
	}
	timeOfProject.Seconds += seconds
	entry.TimesPerProject[projectId] = timeOfProject
}
